In [1]:
#D682 Task - 1

#Rbrune15

# Library and data import

In [3]:
# Importing our libraries and model. Pandas and numpy are used to deal with data and numerical data. Scikit learn is an ML library that provides algorithms and eval tools. 
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_percentage_error

In [4]:
# Loading in data
# We are loading in our provided dataset as a csv. 
data = pd.read_csv("C:/Users/Ryan Brunelle/Desktop/d682-ai-optimization-for-computer-scientists/DQN1 Dataset.csv")

In [5]:
# Checking data loaded in
data.head(10)

,datetimeEpoch,tempmax,tempmin,temp,feelslikemax,feelslikemin,feelslike,pm2.5,no2,co2,...,sunriseEpoch,sunsetEpoch,moonphase,tempRange,heatIndex,severityScore,month,dayOfWeek,isWeekend,healthRiskScore
0,1725692400,106.10,91.00,98.50,104.00,88.10,95.90,38.56,22.62,439.01,...,1725714410,1725759893,0.16,15.10,95.92,4.43,9,2,True,10.52
1,1725681822,91.73,74.13,81.96,89.68,72.87,81.21,27.40,24.64,451.22,...,1725749687,1725842528,0.13,19.19,81.95,4.65,9,2,True,10.70
2,1725632581,73.02,63.37,69.00,72.70,62.75,70.63,32.12,25.46,477.08,...,1725732283,1725762265,0.13,11.26,70.53,3.34,9,2,True,9.85
3,1725692400,85.00,72.90,78.60,87.80,72.90,79.70,36.27,39.94,488.41,...,1725715658,1725761086,0.16,12.10,80.98,3.10,9,2,True,10.34
4,1725716396,103.30,74.19,85.79,102.63,75.29,84.95,29.63,30.12,461.76,...,1725758844,1725743102,0.19,28.36,88.40,2.51,9,2,True,9.93
5,1725624159,73.31,63.97,69.17,72.09,63.24,67.48,37.90,29.32,462.03,...,1725693621,1725710943,0.14,10.35,70.08,3.28,9,2,True,9.72
6,1725731004,106.09,91.64,97.86,105.06,89.37,95.11,32.94,30.67,441.87,...,1725692154,1725760117,0.15,15.25,95.61,4.32,9,2,True,10.47
7,1725663296,62.98,50.70,55.55,64.78,50.56,58.10,11.74,15.59,434.39,...,1725684207,1725713400,0.14,12.16,83.58,4.30,9,2,True,10.81
8,1725685200,88.10,70.00,78.10,90.00,70.00,78.90,35.39,21.96,466.26,...,1725710532,1725755758,0.15,18.10,80.09,5.08,9,2,True,11.47
9,1725630609,105.45,76.50,85.00,102.68,74.42,85.01,36.99,38.63,471.11,...,1725736324,1725744492,0.15,27.47,87.06,2.56,9,2,True,9.89


# After verifying the dataset loaded in correctly we can see it contains all features including the historical air quality, weather conditions, and corresponding health risk scores I need to train the model. 

In [7]:
# Pushing to repo. 

# Now that the data is loaded in, I must identify and define the input features and target variable by setting X to the pm2.5, no2, co2, temp, humidity features of our dataset while setting the Health Risk Score (target variable) to Y. 

In [9]:
# Setting variables (X=Independent, y=Dependent)
X = data[['pm2.5', 'no2', 'co2', 'temp', 'humidity']]
y = data['healthRiskScore']

In [10]:
X.head()

,pm2.5,no2,co2,temp,humidity
0,38.56,22.62,439.01,98.50,21.00
1,27.40,24.64,451.22,81.96,52.84
2,32.12,25.46,477.08,69.00,78.17
3,36.27,39.94,488.41,78.60,72.30
4,29.63,30.12,461.76,85.79,48.21


In [11]:
y.head()

0    10.52
1    10.70
2     9.85
3    10.34
4     9.93
Name: healthRiskScore, dtype: float64

# Next we have to split the data sets (X, y) into two sets. One called train in which is the data the model trains itself on, and test which it evaluates its predictive performance on. This is done using Scikit learns "train_test_split" function. 

In [13]:
# Split data to training and testing sets thru TrainTestSplit
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=28)

Before fitting the model to our split datasets, we have to call and initialize the RandomForestRegressor model from scikits libray. 

In [15]:
# Initiate the Random Forest Regressor model
rf_model = RandomForestRegressor(
    n_estimators=100,
    random_state=28
)

We can now fit the model to our X and y datasets using .fit function.

In [17]:
# Fitting the model to train and test sets
rf_model.fit(X_train, y_train)

RandomForestRegressor(random_state=28)

Now that the model is fit we can generate the predictions. The predictions are made using the unseen testing data. 

In [19]:
# Model predicting
y_pred = rf_model.predict(X_test)

For our evaluation of the results I have chosen to use Mean Squared Erron (rmse) which measures the magnitude of prediction errors and Mean Absolute Percentage Error (mape) which provides a more interpretable percent based error metric. 

In [21]:
# Eval Metrics (mean squared error, mean absolute % error)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
mape = mean_absolute_percentage_error(y_test, y_pred)

In [22]:
# Printing results
print("RMSE:", rmse)

RMSE: 0.26963057161605364


In [23]:
print("MAPE", mape)

MAPE 0.020850959480515963


The RMSE value of 0.2696 indicates that the model’s predictions closely align with observed health risk scores, with relatively small average errors.
The MAPE value of approximately 2.09% suggests that the model is highly accurate, with predictions deviating only slightly from actual values on average.

# TASK 2

In [44]:
# Storing baseline results from Task 1.
baseline_rmse = rmse
baseline_mape = mape


print("Baseline RMSE:", baseline_rmse)
print("Baseline MAPE:", baseline_mape)

Baseline RMSE: 0.26963057161605364
Baseline MAPE: 0.020850959480515963


# B1 - Optimization Technique #1: RandomSearchCV

RandomizedSearchCV improves model performance by testing different hyperparameter combinations to find a strong overall configuration. 

In [48]:
# Import library
from sklearn.model_selection import RandomizedSearchCV

In [64]:
# Configure RandomSearchCV and set parameters
rf_param_dist = {
    "n_estimators": [100, 200, 400, 600],
    "max_depth": [None, 5, 10, 15, 20],
    "min_samples_split": [2, 5, 10, 20],
    "min_samples_leaf": [1, 2, 4, 8],
    "max_features": ["sqrt", "log2", 0.8],
    "bootstrap": [True, False]
}

random_search = RandomizedSearchCV(
    estimator=RandomForestRegressor(random_state=42),
    param_distributions=rf_param_dist,
    n_iter=25,
    scoring="neg_root_mean_squared_error",
    cv=5,
    n_jobs=-1,
    random_state=28
)

In [66]:
# Fit model, define, and print new RandomizedSearch Eval metrics.
random_search.fit(X_train, y_train)

optimized_rf_random = random_search.best_estimator_
y_pred_random = optimized_rf_random.predict(X_test)

rmse_random = np.sqrt(mean_squared_error(y_test, y_pred_random))
mape_random = mean_absolute_percentage_error(y_test, y_pred_random)

print("RandomizedSearch RMSE:", rmse_random)
print("RandomizedSearch MAPE:", mape_random)

RandomizedSearch RMSE: 0.30061211068609983
RandomizedSearch MAPE: 0.022096066487431787


# B1 - Optimization Technique #2: GridSearchCV refinement

GridSearchCV fine tunes the model by exhaustively testing a small, targeted set of hyperparameter values around the best initial results.

In [69]:
# Import Library
from sklearn.model_selection import GridSearchCV

In [70]:
# Define Params and Configure GridSearch
grid_params = {
    "n_estimators": [random_search.best_params_["n_estimators"]],
    "max_depth": [random_search.best_params_["max_depth"]],
    "min_samples_split": [random_search.best_params_["min_samples_split"]],
    "min_samples_leaf": [random_search.best_params_["min_samples_leaf"]],
    "max_features": [random_search.best_params_["max_features"]],
    "bootstrap": [random_search.best_params_["bootstrap"]],
}

grid_search = GridSearchCV(
    estimator=RandomForestRegressor(random_state=42),
    param_grid=grid_params,
    scoring="neg_root_mean_squared_error",
    cv=5,
    n_jobs=-1
)

In [71]:
# Fit model, define, and print new GridsearchCV Eval metrics.
grid_search.fit(X_train, y_train)

optimized_rf_grid = grid_search.best_estimator_
y_pred_grid = optimized_rf_grid.predict(X_test)

rmse_grid = np.sqrt(mean_squared_error(y_test, y_pred_grid))
mape_grid = mean_absolute_percentage_error(y_test, y_pred_grid)

print("GridSearch RMSE:", rmse_grid)
print("GridSearch MAPE:", mape_grid)

GridSearch RMSE: 0.30061211068609983
GridSearch MAPE: 0.022096066487431787


# B2 - Regularization Methods - max_depth & min_samples_leaf

Max_depth limits how deep each decision tree can grow.


Min_samples_leaf sets minimum number of training samples required in a node and requires each leaf to represent more samples.

In [85]:
# Define params and fit with new regularization vars. 
regularized_rf = RandomForestRegressor(
    n_estimators=300,
    random_state=28,
    max_depth=10,
    min_samples_leaf=4
)

regularized_rf.fit(X_train, y_train)

# Evaluate regularized model
y_pred_reg = regularized_rf.predict(X_test)
rmse_reg = np.sqrt(mean_squared_error(y_test, y_pred_reg))
mape_reg = mean_absolute_percentage_error(y_test, y_pred_reg)

print("Regularized RF RMSE:", rmse_reg)
print("Regularized RF MAPE:", mape_reg)

Regularized RF RMSE: 0.28324611493181623
Regularized RF MAPE: 0.022306035838072554


# B3 - Ensemble Technique #1: ExtraTrees

Extra Trees increase randomness during tree splitting,  improving generalization and reducing variance.

In [91]:
# Import Library
from sklearn.ensemble import ExtraTreesRegressor

In [93]:
# Define and establish extra_trees model and params.
extra_trees = ExtraTreesRegressor(
    n_estimators=500,
    random_state=28,
    n_jobs=-1
)

In [97]:
# Fit model
extra_trees.fit(X_train, y_train)

ExtraTreesRegressor(n_estimators=500, n_jobs=-1, random_state=28)

In [99]:
# Evaluate Extra Trees model and print.
y_pred_et = extra_trees.predict(X_test)
rmse_et = np.sqrt(mean_squared_error(y_test, y_pred_et))
mape_et = mean_absolute_percentage_error(y_test, y_pred_et)

print("ExtraTrees RMSE:", rmse_et)
print("ExtraTrees MAPE:", mape_et)

ExtraTrees RMSE: 0.25872225042697966
ExtraTrees MAPE: 0.020570233083371612


# B3 - Ensemble Technique #2: Gradient Boosting

Gradient Boosting builds models sequentially with each new model correcting errors from previous one.

In [105]:
# Import Library
from sklearn.ensemble import GradientBoostingRegressor

In [107]:
# Define GRboost
gbr = GradientBoostingRegressor(random_state=42)

In [111]:
# Fit model
gbr.fit(X_train, y_train)

GradientBoostingRegressor(random_state=42)

In [113]:
# Evaluate Gradient Boosting model and print.
y_pred_gbr = gbr.predict(X_test)
rmse_gbr = np.sqrt(mean_squared_error(y_test, y_pred_gbr))
mape_gbr = mean_absolute_percentage_error(y_test, y_pred_gbr)

print("GradientBoosting RMSE:", rmse_gbr)
print("GradientBoosting MAPE:", mape_gbr)

GradientBoosting RMSE: 0.2731328215543034
GradientBoosting MAPE: 0.022017421034752765


# C2 - Comparison Table of all results vs Baseline

In [117]:
# This table summarizes performance improvements across all optimization, regularization, and ensemble techniques.

results = pd.DataFrame([
    ["Baseline RF (Task 1)", baseline_rmse, baseline_mape],
    ["RF RandomizedSearch", rmse_random, mape_random],
    ["RF GridSearch", rmse_grid, mape_grid],
    ["Regularized RF", rmse_reg, mape_reg],
    ["ExtraTrees", rmse_et, mape_et],
    ["GradientBoosting", rmse_gbr, mape_gbr],
], columns=["Model", "RMSE", "MAPE"]).sort_values("RMSE")

results

,Model,RMSE,MAPE
4,ExtraTrees,0.258722,0.020570
0,Baseline RF (Task 1),0.269631,0.020851
5,GradientBoosting,0.273133,0.022017
3,Regularized RF,0.283246,0.022306
1,RF RandomizedSearch,0.300612,0.022096
2,RF GridSearch,0.300612,0.022096


# See "Task2report" in repo for summary and explanation of results and completed task parts.